hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordMerger.java

```diff
@@ -30,9 +34,19 @@
  * It can implement the merging logic of HoodieRecord of different engines
  * and avoid the performance consumption caused by the serialization/deserialization of Avro payload.
  */
-public interface HoodieMerge extends Serializable {
-  
-  HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer);
+@PublicAPIClass(maturity = ApiMaturityLevel.EVOLVING)
+public interface HoodieRecordMerger extends Serializable {
+
+  /**
+   * This method converges combineAndGetUpdateValue and precombine from HoodiePayload.
+   * It'd be associative operation: f(a, f(b, c)) = f(f(a, b), c) (which we can translate as having 3 versions A, B, C
+   * of the single record, both orders of operations applications have to yield the same result)
+   */
+  Option<HoodieRecord> merge(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) throws IOException;
```

Comment:
For inserts, deletes - we will not be calling merge()? i.e `older` and `newer` will always be non-null? Something to think about is - whether we would want the user to be able to get call backs for those as well

Reply:
Maybe insert a marker in HoodieRecord?

Reply:
@vinothchandar not sure i understand the original question? We will retain the same semantic w/ tombstone value of `HoodieRecord` being passed around for deletion

Reply:
@alexeykudinkin Maybe vc means merge function will deduplicate the records for insertion.  Do you think we should put the shouldCombine marker in the record? cc @vinothchandar

Reply:
[sorry been out of GH with some fam stuff I have been dealing with in India.]

yes wonder if we should make this API more generic and provide ways for user to transform if needed for inserts/deletes as well. (there could be use-cases today where some additional metadata is computed here, e.g arrival time etc.). Users who currently wrote some custom code inside `HoodieRecordPaylod::getInsertValue()` have no way of translating into the new API right

 Default can just return null/empty for deletes and newer as-is for inserts. I do feel it makes for a more streamlined experience.